### PR TITLE
Add deterministic train_test_split

### PR DESCRIPTION
**Pull request by FelixAbrahamsson**
Deterministic `train_test_split` function that will always split datapoints the same way given a fixed seed. At the moment the interface is limited, it accepts the following arguments
* `df` - a pandas dataframe
* `key` - a string which must be a column in `df`
* `test_size` (optional) - float, same as sklearn version
* `random_state` (optional) - integer, same as sklearn version
* `stratify` (optional) - a string (column in `df`) or an iterable of strings (columns in `df`)

### CODE REVIEW

File: workflow/train_test_split.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        test_df = multilabel_stratified_split(df, key, stratify, test_size)
+
+    train_df = df[~df[key].isin(test_df[key])]
+    np.random.set_state(original_random_state)
+
+    return train_df, test_df
+
+
+def get_test_split(df, key, test_size, resolution=100000):
+    salt = np.random.randint(0, resolution)
```
Comment (samedii) [inline workflow/train_test_split.py]: I don't think we need the random salt anymore? We can salt with the seed since we are hashing again?

File: workflow/train_test_split.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+def train_test_split(df, key, test_size=0.2, random_state=None, stratify=None):
+
+    original_random_state = np.random.get_state()
+    np.random.seed(random_state)
+
+    if stratify is None:
+        test_df = get_test_split(df, key, test_size)
+    else:
+        test_df = multilabel_stratified_split(df, key, stratify, test_size)
```
Comment (samedii) [inline workflow/train_test_split.py]: I dislike the change in behaviour from the classic train_test_split that does not support multilabel. If we want to support  multilabel here I think we need another argument to the function

Also, classic stratify should then raise Exception as not supported yet imo.

File: workflow/train_test_split.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    train_df = df[~df[key].isin(test_df[key])]
+    np.random.set_state(original_random_state)
+
+    return train_df, test_df
+
+
+def get_test_split(df, key, test_size, resolution=100000):
+    salt = np.random.randint(0, resolution)
+    split_key = df[key].apply(lambda x: ((hash(hash(x) + salt) % resolution)))
+    return df[split_key < resolution * test_size]
```
Comment (samedii) [inline workflow/train_test_split.py]: Something like this?
```
hash = df[key].apply(lambda x: ((hash(hash(x) + salt) % resolution)))
is_test = (split_key < resolution * test_size)
return df[~is_test], df[is_test]
``` 
Or just return `is_test`

Would simplify the code in `train_test_split`

File: workflow/train_test_split.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
+from collections.abc import Iterable
+import numpy as np
+import pandas as pd
+import pytest
+
+
+def train_test_split(df, key, test_size=0.2, random_state=None, stratify=None):
```
Comment (samedii) [inline workflow/train_test_split.py]: `key` to `unique_key`?

File: workflow/train_test_split.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    return train_df, test_df
+
+
+def get_test_split(df, key, test_size, resolution=100000):
+    salt = np.random.randint(0, resolution)
+    split_key = df[key].apply(lambda x: ((hash(hash(x) + salt) % resolution)))
+    return df[split_key < resolution * test_size]
+
+
+def multilabel_stratified_split(df, key, stratify, test_size):
```
Comment (samedii) [inline workflow/train_test_split.py]: This looks like the implementation that I talked about before but it should be weaker than the old one? Am I missing something?

File: workflow/train_test_split.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+def get_labels(df, stratify):
+
+    if type(stratify) is str:
+        return sorted(df[stratify].unique())
+    else:
+        if not isinstance(stratify, Iterable):
+            raise ValueError(
+                f'Argument stratify must be str or iterable, was {stratify}'
+            )
+        return sorted(stratify)
```
Comment (samedii) [inline workflow/train_test_split.py:55]: What is this other type of stratify you are supporting? When would you use it?

**Review by samedii — CHANGES_REQUESTED** [review]
Looks nice. I had some comments and minor RCs. Impressive tests :O did not look through them properly yet though

File: workflow/train_test_split.py (hunk 2 above)
Comment (FelixAbrahamsson) [reply workflow/train_test_split.py]: This is a problem with confusing naming I think. The function `multilabel_stratified_split` actually performs the "classic stratify" in the case that i receives a string for the stratify parameter. Perhaps it should just be named `stratified_split`?

File: workflow/train_test_split.py (hunk 4 above)
Comment (FelixAbrahamsson) [reply workflow/train_test_split.py]: Agreed, changing :)

File: workflow/train_test_split.py (hunk 6 above)
Comment (FelixAbrahamsson) [reply workflow/train_test_split.py:55]: I replied above but can clarify. If stratify is type str it will do a normal stratify, if it is an iterable it will do a multilabel stratification. Since the algorithm works the same way for both cases I let it sit in the same function, but maybe it should be split up?

File: workflow/train_test_split.py (hunk 1 above)
Comment (FelixAbrahamsson) [reply workflow/train_test_split.py]: True! It would require passing around the seed, but still maybe a better solution?

**Comment by samedii** [comment]
Me and Felix talked and realized that multilabel split like this does not work well for the imbalanced overlapping case. Decided to refactor to a more specific solution for the non-multilabel case.

File: workflow/train_test_split.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        test_df = stratified_split(df, unique_key, stratify, test_size, seed)
+
+    train_df = df[~df[unique_key].isin(test_df[unique_key])]
+
+    return train_df, test_df
+
+
+def get_test_split(df, unique_key, test_size, seed=0, resolution=100000):
+    split_unique_key = df[unique_key].apply(
+        lambda x: ((hash(hash(x) + seed) % resolution))
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:26]:
```suggestion
        lambda x: hash(hash(x) ^ seed) % resolution
```

File: workflow/train_test_split.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+import numpy as np
+import pandas as pd
+import pytest
+
+
+def train_test_split(
+    df,
+    unique_key,
+    test_size=0.2,
+    seed=0,
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:11]:
```suggestion
    seed=hash("ml-workflow"),
```

File: workflow/train_test_split.py (hunk 7 above)
Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:26]: ``` ^ hash("something")```
is to make the output have the right distribution even if x does not
for example `hash(x) == x` for integers would make `hash(hash(x) + 0) == x` and all bets are off :)
better safe then sorry, don't want to get bad results due to some bad splitting :)

Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:26]: `^` (xor) preserves entropy which is a good property for hashing
combining multiple elements into the same hash value can be a bit tricky, but we only have one variable element and a constant one.
for multiple elements you have to watch out that for example the hash is not the same when elements are reordered or that "ab" "c" and "a" "bc" arent hashed to the same value :)

File: workflow/train_test_split.py (hunk 6 above)
Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:55]: I think its clear from the code :) perhaps  something that could be made better later, but its most definitely good enough to pull into master

File: workflow/train_test_split.py (hunk 8 above)
Comment (samedii) [reply workflow/train_test_split.py:11]: Motivation ?

File: workflow/train_test_split.py (hunk 6 above)
Comment (samedii) [reply workflow/train_test_split.py:55]: Why would you want to input a list as stratify?

Comment (samedii) [reply workflow/train_test_split.py:55]: Ah it's when you have one column per label in the dataframe

File: workflow/train_test_split.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    )
+    return df[split_unique_key < resolution * test_size]
+
+
+def stratified_split(df, unique_key, stratify, test_size, seed):
+
+    test_dfs = []
+    for label in get_labels(df, stratify):
+        if type(stratify) is str:
+            df_part = df[df[stratify] == label]
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:36]:
```suggestion
            df_accepted = df[df[stratify] == label]
```
or something a bit more descriptive to make it easier to follow

File: workflow/train_test_split.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def get_test_split(df, unique_key, test_size, seed=0, resolution=100000):
+    split_unique_key = df[unique_key].apply(
+        lambda x: ((hash(hash(x) + seed) % resolution))
+    )
+    return df[split_unique_key < resolution * test_size]
+
+
+def stratified_split(df, unique_key, stratify, test_size, seed):
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:31]:
```suggestion
def stratified_split(df, unique_key, stratify, test_size, seed=0):
```

File: workflow/train_test_split.py (hunk 9 above)
Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:36]: `accepted` borrowed from https://en.wikipedia.org/wiki/Rejection_sampling which is perhaps not the best

File: workflow/train_test_split.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+    test_dfs = []
+    for label in get_labels(df, stratify):
+        if type(stratify) is str:
+            df_part = df[df[stratify] == label]
+        else:
+            df_part = df[df[label].notnull()]
+        test_dfs.append(get_test_split(df_part, unique_key, test_size, seed))
+        df = df[~df[unique_key].isin(df_part[unique_key])]
+    assert len(df) == 0, f'Unable to stratify on {stratify}'
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:41]: tried to reason about this 3 times :P
what is the minimum number of labels that can end up in test split without it asserting?

File: workflow/train_test_split.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        seed=1,
+        stratify='multiclass',
+    )
+    for label in test_df['multiclass'].unique():
+        ratio = (
+            len(test[test['multiclass'] == label])
+            / len(test_df[test_df['multiclass'] == label])
+        )
+        if ratio < 0.1 or ratio > 0.3:
+            raise Exception(f'Expected ratio close to 0.2, got {ratio}')
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:103]: duplicate

File: workflow/train_test_split.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        assert ratio < 0.3 and ratio > 0.1
```
Comment (Jim-Holmstroem) [inline workflow/train_test_split.py:104]:
```suggestion
        assert 0.1 < ratio < 0.3
```
(actually works as expected in python)

File: workflow/train_test_split.py (hunk 8 above)
Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:11]: 0 is kinda special in that it is a unit for many operations (addition, xor etc)
so for example `hash(hash(34) ^ 0) == 34` which is kinda scary
this removes that scary

File: workflow/train_test_split.py (hunk 13 above)
Comment (Jim-Holmstroem) [reply workflow/train_test_split.py:104]: https://github.com/codewell/ml-workflow/pull/83

**Review by Jim-Holmstroem — CHANGES_REQUESTED** [review]
I was happy reading this PR, well made! :D

**Comment by samedii** [comment]
I realized that the multilabel behaviour can be recreated with a created single column. You would just keep some arbitrary (the same, determinstic) column in each case where we have overlap

**Comment by samedii** [comment]
Let's pick this up on monday and do the last work to merge?

**Comment by samedii** [comment]
Cleaning and closing this. We can reopen if we want this functionality later
